Added support for creating ECC asymmetric keys

fmonteiro93 opened the pull request:
Features:
- Added support to generate ECC keys with all the OpenSSL NIST curves.
- Introduced guards to functionality that only supports RSA keys
- Extended unit tests

fmonteiro93 committed:
Ecc asymmetric key support

First working version of the ECC key generator
(WIP)

fmonteiro93 committed:
Added get type method to asym public/privatekey

fmonteiro93 committed:
Added default generate method for ECC keys

fmonteiro93 committed:
Added ECC unit tests

fmonteiro93 committed:
Added mocks for new exposed Openssl functions.

fmonteiro93 committed:
Added new interface, guards and unit tests

Added getCurve interface, extended ECC key unit test and added
guards to the functionality using RSA keys

pmvfonseca commented on src/key.cpp:
```diff
@@ -31,7 +32,18 @@ using namespace openssl;
 AsymmetricKeypair AsymmetricKeypair::generate()
```
Should the name be changed to generateRSA?

fmonteiro93 replied:
I don't think it's a good idea to change the name of the method since it will break backwards compatibility. Maybe we could create a new one called generateRSA and mark this one as deprecated.

pmvfonseca replied:
You're right, that's sounds like a good solution.

pmvfonseca commented on src/key.cpp:
```diff
@@ -84,4 +96,36 @@ AsymmetricKey RSASpec::generate() const
+    if(getType() != KeyTypes::ECC){
+        throw MoCOCrWException("Functionality only supported for ECC keys");
+    }
+    const EC_GROUP *group = _EC_KEY_get0_group(_key->pkey.ec);
```
Missing exception handling?

fmonteiro93 replied:
Agreed.

pmvfonseca commented on src/mococrw/key.h:
```diff
@@ -77,12 +66,43 @@ class AsymmetricPublicKey
 
+    /**
+     * Converts the asymmetric public key to the PKCS8 format that can be written in a PEM file.
+     * @return public keu in PKCS format
```
keu -> key

:D

fmonteiro93 replied:
:+1:

pmvfonseca commented on src/mococrw/padding_mode.h:
```diff
@@ -79,6 +80,9 @@ class NoPadding: public RSAPadding {
     */
     int getDataMaxSize(const AsymmetricPublicKey& key) const override
     {
+        if(key.getType() != AsymmetricKey::KeyTypes::RSA){
```
I wonder if having two derived classes, (AsymmetricKeyRSA and AssymetricKeyEC, for example) would be better in the sense that having these guards would no longer be necessary

fmonteiro93 replied:
I don't think we can do this cleanly without breaking backwards compatibility. From my point of view the idea of the AsymmetricKey class is to abstract the underlying algorithms used, especially when loading a key from a PEM string

pmvfonseca replied:
Backwards compatibility is indeed a problem, but these kind of guards are a bit inelegant in my opinion.

fmonteiro93 replied:
Most of the guards will be removed once we add support for ECC signatures and certificates.
Only the encryption functionality will need them

fmonteiro93 replied:
Removed some of the guards by replacing RSA_size with EVP_PKEY_size

fmonteiro93 committed:
Changes after Code Review

fmonteiro93 committed:
Deprecated generic key generate() method

fmonteiro93 committed:
Added EVP_PKEY_size and fixed Oaep padding

- Removed usage of RSA_size for EVP_PKEY_size().
- Fixed issue on oaep padding getMaxSize method
- Extended unit tests

fmonteiro93 committed:
Changes after code review
